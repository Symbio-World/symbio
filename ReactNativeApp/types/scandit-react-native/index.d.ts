declare module 'scandit-react-native'
