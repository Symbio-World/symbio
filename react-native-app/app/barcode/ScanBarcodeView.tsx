import * as React from 'react'
import {
  AppState,
  AppStateStatus,
  View,
  Platform,
  PermissionsAndroid,
  BackHandler,
  Text,
} from 'react-native'
import { BarcodePicker } from 'scandit-react-native'
import { t } from 'react-native-tailwindcss'
import { settings } from './ScanBarcodeViewSettings'
import { Scanner, Session } from './Scanner'
import { useNavigation } from '@react-navigation/native'
import { BigRoundedButton } from '../ui-kit/BigRoundedButton'
import { Icons } from '../ui-kit/Icon'
import Icon from 'react-native-vector-icons/MaterialIcons';

type Props = {
  onScan?: (barcode: string) => void
  isActive?: boolean

  // a hack to test scanner ref as createNodeMock is not working https://github.com/callstack/react-native-testing-library/issues/227
  useRef?: typeof React.useRef
}

const isAndroidMarshmallowOrNewer = () => {
  return Platform.OS === 'android' && Platform.Version >= 23
}

const hasCameraPermission = () =>
  isAndroidMarshmallowOrNewer()
    ? PermissionsAndroid.check(PermissionsAndroid.PERMISSIONS.CAMERA)
        .then((isGranted) => isGranted)
        .catch(() => false)
    : Promise.resolve(true)

const requestCameraPermission = () => {
  return new Promise((resolve, reject) => {
    if (isAndroidMarshmallowOrNewer()) {
      PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.CAMERA)
        .then((granted) => {
          if (granted === PermissionsAndroid.RESULTS.GRANTED) {
            console.log('Android Camera Permission has been granted.')
            return resolve()
          }
          console.log(
            'Android Camera Permission has been denied - the app will shut itself down.',
          )
          return reject()
        })
        .catch((err) => {
          console.warn(err)
          reject()
        })
      return
    }
    resolve()
  })
}

const checkForCameraPermission = () => {
  return new Promise((resolve, reject) => {
    hasCameraPermission()
      .then((hasPermission) => {
        if (hasPermission) {
          resolve()
          return
        }
        requestCameraPermission().then(resolve).catch(reject)
      })
      .catch(reject)
  })
}

export const ScanBarcodeView: React.FC<Props> = ({
  onScan = () => {},
  isActive = true,
  useRef = React.useRef,
}) => {
  const navigation = useNavigation()
  const scanner = useRef<Scanner>(null)
  const [isScannerStarted, setScannerStarted] = React.useState(false)

  const startScanning = () => {
    try {
      scanner.current?.startScanning()
      setScannerStarted(true)
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }

  const stopScanning = () => {
    try {
      scanner.current?.stopScanning()
      setScannerStarted(false)
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }

  const resumeScanning = () => {
    try {
      scanner.current?.resumeScanning()
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }
  const pauseScanning = () => {
    try {
      scanner.current?.pauseScanning()
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }

  const disableTorch = () => {
    try {
      scanner.current?.setTorchEnabled(false)
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }

  React.useEffect(() => {
    const handleAppStateChange = (nextAppState: AppStateStatus) => {
      if (nextAppState.match(/inactive|background/)) {
        pauseScanning()
      } else {
        checkForCameraPermission()
          .then(startScanning)
          .catch(BackHandler.exitApp)
      }
    }

    AppState.addEventListener('change', handleAppStateChange)
    checkForCameraPermission().then(startScanning).catch(BackHandler.exitApp)
    return () => {
      AppState.removeEventListener('change', handleAppStateChange)
      stopScanning()
    }
  }, [navigation])

  /* eslint-disable react-hooks/exhaustive-deps */
  React.useEffect(() => {
    if (!isScannerStarted) return
    if (isActive) {
      resumeScanning()
      return
    }
    pauseScanning()
  }, [isActive])
  /* eslint-enable react-hooks/exhaustive-deps */

  React.useEffect(() => {
    disableTorch()
  }, [])

  const handleScan = (session: Session) => {
    const barcode = session.newlyRecognizedCodes[0].data
    onScan(barcode)
  }

  return (
    <View style={[t.flex1]}>
      <BarcodePicker
        onScan={handleScan}
        scanSettings={settings}
        ref={scanner}
        style={[t.flex1]}
      />
      <View
        style={[
          t.absolute,
          t.top0,
          t.wFull,
          t.h16,
          t.justifyCenter,
          t.bgPrimary,
        ]}
      >
        <Text style={[t.textWhite, t.textCenter, t.fontBold, t.text3xl]}>
          Scan barcode
        </Text>
      </View>
      <View
        style={[t.absolute, t.bottom0, t.wFull, t.flexRow, t.justifyEvenly, t.mB6]}
      >
        <BigRoundedButton
          onPress={() => {}}
          title="flashlight"
          variant="outline"
          icon={<Icon name="flashlight" size={24} color="#900" />}
        />
        <BigRoundedButton
          onPress={() => {}}
          title="history"
          variant="fill"
          icon={<Icon name="history" size={24} color="#900" />}
        />

      </View>
    </View>
  )
}
