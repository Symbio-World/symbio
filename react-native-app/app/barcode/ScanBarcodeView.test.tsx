import * as React from 'react'
import { render, fireEvent } from 'react-native-testing-library'
import { BarcodePicker } from 'scandit-react-native'
import { ScanBarcodeView } from './ScanBarcodeView'
import * as fixture from './ScanBarcodeView.fixture'

const mockUseRef = (obj: unknown) => () =>
  Object.defineProperty({}, 'current', {
    get: () => obj,
    set: () => {},
  })

describe('ScanBarcodeView', () => {
  const navigation: any = {
    navigate: jest.fn(),
  }

  it('renders correctly', () => {
    const { toJSON } = render(<ScanBarcodeView navigation={navigation} />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('triggers onScan prop', () => {
    const handleScan = jest.fn()
    const useRef = mockUseRef({ startScanning: jest.fn() })
    const { getByType } = render(
      <ScanBarcodeView
        onScan={handleScan}
        useRef={useRef}
        navigation={navigation}
      />,
    )
    fireEvent(getByType(BarcodePicker), 'onScan', fixture.session)
    expect(handleScan).toHaveBeenCalledWith(fixture.barcode)
  })
})
