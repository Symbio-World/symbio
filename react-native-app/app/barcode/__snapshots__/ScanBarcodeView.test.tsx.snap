// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanBarcodeView renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
    ]
  }
>
  <BarcodePicker
    onBarcodeFrameAvailable={[Function]}
    onChangeTrackedCodes={[Function]}
    onPropertyChanged={[Function]}
    onRecognizeNewCodes={[Function]}
    onScan={[Function]}
    onSettingsApplied={[Function]}
    onTextRecognized={[Function]}
    onWarnings={[Function]}
    scanSettings={
      ScanSettings {
        "symbologies": Object {
          "code128": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "code39": SymbologySettings {
            "activeSymbolCounts": Array [
              7,
              8,
              9,
              10,
              11,
              12,
              13,
              14,
              15,
              16,
              17,
              18,
              19,
              20,
            ],
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "data-matrix": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "ean13": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "ean8": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "itf": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "qr": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "upca": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "upce": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
        },
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "position": "absolute",
        },
        Object {
          "top": 0,
        },
        Object {
          "width": "100%",
        },
        Object {
          "height": 64,
        },
        Object {
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "#B35340",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
          },
          Object {
            "textAlign": "center",
          },
          Object {
            "fontWeight": "bold",
          },
          Object {
            "fontSize": 30,
          },
        ]
      }
    >
      Scan barcode
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "position": "absolute",
        },
        Object {
          "bottom": 0,
        },
        Object {
          "width": "100%",
        },
        Object {
          "flexDirection": "row",
        },
        Object {
          "justifyContent": "space-evenly",
        },
        Object {
          "marginBottom": 24,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "borderWidth": 2,
            },
            Object {
              "borderColor": "#B35340",
            },
            Object {
              "width": 96,
            },
            Object {
              "height": 96,
            },
            Object {
              "borderRadius": 9999,
            },
            Object {
              "flexDirection": "column",
            },
            Object {
              "alignItems": "center",
            },
            Object {
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "width": 24,
              },
              Object {
                "height": 24,
              },
              Object {
                "marginBottom": 8,
              },
            ]
          }
        >
          <SvgMock
            height="100%"
            width="100%"
          />
        </View>
        <View>
          <Text
            style={
              Array [
                Object {
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 16,
                },
                Object {
                  "fontWeight": "bold",
                },
                Object {
                  "color": "#B35340",
                },
              ]
            }
          >
            flashlight
          </Text>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#B35340",
            },
            Object {
              "width": 96,
            },
            Object {
              "height": 96,
            },
            Object {
              "borderRadius": 9999,
            },
            Object {
              "flexDirection": "column",
            },
            Object {
              "alignItems": "center",
            },
            Object {
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "width": 24,
              },
              Object {
                "height": 24,
              },
              Object {
                "marginBottom": 8,
              },
            ]
          }
        >
          <SvgMock
            height="100%"
            width="100%"
          />
        </View>
        <View>
          <Text
            style={
              Array [
                Object {
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 16,
                },
                Object {
                  "fontWeight": "bold",
                },
                Object {
                  "color": "#fff",
                },
              ]
            }
          >
            history
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
